tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules2/StratifiedRule.scala

```diff
@@ -0,0 +1,23 @@
+package at.forsyte.apalache.tla.bmcmt.rules2
+
+import at.forsyte.apalache.tla.bmcmt.ArenaCell
+import at.forsyte.apalache.tla.lir.TlaEx
+
+/**
+ * @author
+ *   Jure Kukovec
+ */
+trait StratifiedRule[T] {
+  def isApplicable(ex: TlaEx, scope: RewriterScope): Boolean
```

Comment:
The thing is, the rewriter uses `isApplicable` to determine which rule should fire, and it's best that be kept lightweight. I think it's possible to do it the way you suggest, and my intuition says it shouldn't break, but it's probably a bit less trivial than just removing the method.

Reply:
I've made a separate issue for this (https://github.com/informalsystems/apalache/issues/2558), because I think it's worthwhile, it would just be a bit too much hassle to implement it as part of this PR, because it'd likely break some of my uncommited work, that's built on top of this.